### ISSUE TITLE
Phase 2.x — v0.4.0: recover coarse F1, hit issue #57 fine-label floors, harden the dual-loss training family

### ISSUE THREAD

**Issue by GirlBossRush** [isp-nexus/mailwoman#116]
Direct follow-up to PR #115 (v0.3.0 → v3.0.0 ship). v0.3.0 landed Stage 2 label expansion + CRF + corpus-v0.3.0, with two known issues that v0.4.0 targets head-on:

1. **Coarse F1 regressed substantially.** Per golden v0.1.2 (4,535 entries): region 0.83 → 0.18, locality 0.65 → 0.27, postcode 0.86 → 0.76 vs v0.2.0's small-slice eval. The new 21-label space pulls per-token softmax mass off the coarse predictions, and step-1800 was nowhere near converged — the ship was a late-best window before recurring divergence, not a peak.
2. **The dual loss (CE + 0.05·CRF NLL) is stability-fragile.** All four v0.3.0 hparam iterations diverged eventually; the only difference was when. CRF NLL is per-sequence + unbounded vs CE per-token + log-bounded, so hand-weighting + LR tuning + grad clipping is doing all the work and there's no margin.

Issue #57's fine-label floors were also not met: venue 0.39 (target 0.60), street 0.27 (target 0.70), house_number 0.78 (target 0.80 — closest).

### 1. Per-token CRF NLL normalization

Replace per-sequence-mean CRF NLL with per-token-normalized form so the loss landscape is self-balancing:

```
crf_loss = -crf.log_likelihood(emissions, tags, mask) / mask.sum()
loss = ce_loss + crf_loss   # no crf_loss_weight needed
```

Eliminates the four-iteration hand-weighting search v0.3.0 went through. Standard NER+CRF practice (AllenNLP defaults to per-token-normalized; FLAIR similar). Defensive `mask.sum().clamp(min=1)` to avoid div-by-zero on empty sequences. Likely re-stabilizes lr=5e-4 + label_smoothing=0.1 (the v0.2.0 baseline minus CRF) since the dual gradients are now comparable scales.

### 2. Recover coarse F1 — train longer

step-1800 is **3.6% of the planned 50K**. The v0.1.0 ship at step-6000 was already a year ahead in convergence depth. v0.4.0 should:

- Run to step 5,000+ minimum before judging val_macro_f1
- If divergence still recurs at step ~2,000 across the new (#1) loss formulation: bisect on hparams; otherwise the per-token CRF norm + 5e-4 LR likely runs cleanly to 50K
- Eval every 250 steps from the start (already configured) so we can early-stop on val_macro_f1 plateau if it's a real plateau rather than a single-step dip

### 3. CE-loss class reweighting toward coarse classes

Compensate for the 21-label space dilution. The class weights aren't free parameters — derive them from the corpus's per-class token frequency, capped:

```
class_weights = (1 / class_freq) ** 0.5  # square root to soften extremes
class_weights[fine_classes] *= 0.5        # halve fine weights so coarse re-dominates
ce_loss = F.cross_entropy(..., weight=class_weights)
```

The 0.5 multiplier is a knob; bias toward coarse since that's the regression to recover. If venue/street/house_number F1 stagnate at the v0.3.0 numbers, dial the multiplier back up.

### 4. Targeted source-weight rebalance

`corpus-v0.3.0`'s NAD adapter (57.9M aligned rows; full venue+street+house_number coverage) pushed the corpus into a fine-label-heavy mix. `usgov-nad: 2.0` source weight may have amplified this. v0.4.0 candidates:

- Drop `usgov-nad` weight to 1.0 (let it be 1× sampled instead of 2×)
- Bump wof-admin + wof-postalcode back up to 2.0–3.0 to re-prioritize coarse-only rows
- Keep ban (3.0) + tiger (4.0) where they are — those carry both fine *and* coarse signal

### 5. JS-side runtime improvements

Two from the i57 postmortem:

- **Load BIO label vocab from `model-card.json` at runtime** instead of the current hardcoded compile-time map. v0.3.0's slice 4 caught this just before publish — the JS classifier originally defaulted to STAGE1_BIO_LABELS and would have silently misrouted venue/street/house_number argmaxes. Stopgap fix landed but is brittle.
- **Viterbi decode in `@mailwoman/neural`.** Eval used CRF Viterbi (that's where the Saint Petersburg fix comes from). The runtime classifier still does per-token argmax. The "Saint Petersburg cannot be clipped" promise is only half-fulfilled — eval-vs-production drift on orphan-I cases. Needs the trained transition matrix exported in the ONNX bundle + a small JS Viterbi loop.

### 6. Out of scope (defer to v0.5.0+)

- Stage 3 organization/POI venue disambiguation
- Top-k Resolver output
- Tokenizer retraining (v0.1.0 tokenizer is fine through Stage 2; revisit if OOV climbs)
- New corpus sources beyond what's already in `corpus/src/adapters/` (the corpus-v0.3.0 → v0.4.0 retrain can reuse the existing build)

## Success metric (per the reframed Phase 2 plan)

Not "hit all the floors" but: **clear progress on at least two of {coarse F1, fine F1, calibration, training stability}** vs v0.3.0.

Specific aspirational targets (issue #57's framing, this iteration's stretch):

- region ≥ 0.6, locality ≥ 0.5, postcode ≥ 0.8 (the coarse regression recovery)
- venue ≥ 0.5, street ≥ 0.4, house_number holds ≥ 0.78 (incremental fine-label progress)
- mean confidence stays calibrated: conf>0.9 bucket accuracy ≥ 0.75
- training runs to step 10,000+ without divergence on the per-token-normalized loss family

If after the changes in §1 + §2 the training is still divergence-prone at step ~2K, that's a v0.4.1 corpus-side investigation (corpus-v0.3.0 has 677M rows — there may be a small high-loss adapter slice that's amplifying CRF gradient spikes).

## Wall-clock estimate

3-5 days:

- Day 1: per-token CRF norm + class-weighted CE + class-frequency derivation (§1, §3 — small Python edits + a smoke train to verify dual gradients comparable scales)
- Day 2: source-weight rebalance in stage2.yaml + smoke (§4)
- Day 3: full 50K training run (~6-8h GPU); eval; ship_v0_4_0.sh
- Day 4: JS-side Viterbi + model-card-loaded label vocab (§5); demo regression test
- Day 5: PR review + npm publish

`corpus-v0.3.0` already exists at `/data/corpus/versioned/v0.3.0/`; no corpus rebuild needed unless source weights change is significant.

## Related

- #57 — v0.3.0 ship (direct predecessor)
- PR #115 — v0.3.0 → v3.0.0 implementation
- DECISIONS.md "v0.3.0 Stage 2 dual loss" — the iteration trail
- `LOG.md` 2026-05-22 entries — what shipped
- Phase 2 plan `docs/articles/plan/phases/PHASE_2_training.md` iteration log

## Disclosure

Issue drafted by Claude (Opus 4.7) post-v3.0.0 docs sweep, working from the i57 container's postmortem ("Stage 2 training is fundamentally unstable on this hparam family") + the v0.3.0 eval report ("coarse F1 regressed substantially") + i57's specific recommendations (per-token CRF norm, JS vocab from model-card.json).
